### PR TITLE
Rescore collapsed documents

### PR DESCRIPTION
**Pull request by fred84**
Add support for rescoring collapsed docs (#27243). Documents at first get collapsed and then rescored.

@jimczi please take a look

### CODE REVIEW

**Comment by jimczi** [comment]
@elasticmachine ok to test

**Review by jimczi — APPROVED** [review]
It looks good to me.
I'll merge if the build passes with the changes, thanks @fred84 !

File: server/src/test/java/org/elasticsearch/search/functionscore/QueryRescorerIT.java [hunk 1]
```diff
@@ -748,4 +754,55 @@ public void testRescorePhaseWithInvalidSort() throws Exception {
...
+        indexDocument(2, "one one value", "a");
+        indexDocument(3, "one one two value", "b");
+        // should be highest on rescore, but filtered out during collapse
+        indexDocument(4, "one two two value", "b");
+
+        refresh("test");
+
+        SearchResponse searchResponse = client().prepareSearch("test")
+            .setTypes("type1")
+            .setQuery(new MatchQueryBuilder("name", "one"))
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/search/functionscore/QueryRescorerIT.java]: The score of this query depends on the number of shards, the default similarity, ... To make sure that we have consistent scoring you can use a `function_score` query like the following:
````
 QueryBuilder query = functionScoreQuery(
            termQuery("name", "one"),
            ScoreFunctionBuilders.fieldValueFactorFunction("my_static_doc_score") 
        ).boostMode(CombineFunction.REPLACE);
````
... and add the `my_static_doc_score` at indexing time.

File: server/src/test/java/org/elasticsearch/search/functionscore/QueryRescorerIT.java [hunk 2]
```diff
@@ -748,4 +754,55 @@ public void testRescorePhaseWithInvalidSort() throws Exception {
...
+            .addRescorer(new QueryRescorerBuilder(new MatchQueryBuilder("name", "two")))
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/search/functionscore/QueryRescorerIT.java]: You can use the same for the rescore with another field for instance

**Review by jimczi — CHANGES_REQUESTED** [review]
Thanks @fred84, you can remove the failing tests, it is no longer needed. Though I left some comments regarding the IT test. I think it needs to be changed to ensure that scoring and ordering are consistent.

**Comment by fred84** [comment]
@jimczi Thanks for reviewing. I'll update PR next week.

File: server/src/test/java/org/elasticsearch/search/functionscore/QueryRescorerIT.java (hunk 1 above)
Comment (fred84) [reply server/src/test/java/org/elasticsearch/search/functionscore/QueryRescorerIT.java]: fixed

File: server/src/test/java/org/elasticsearch/search/functionscore/QueryRescorerIT.java (hunk 2 above)
Comment (fred84) [reply server/src/test/java/org/elasticsearch/search/functionscore/QueryRescorerIT.java]: fixed

**Comment by fred84** [comment]
@jimczi PR updated, now integration test use static scoring.

**Comment by jimczi** [comment]
@elasticmachine ok to test

**Review by jimczi — APPROVED** [review]
Thanks @fred84

**Comment by jimczi** [comment]
I had to revert this change since it doesn't work as expected. I forgot that the collapsed values would also need to be resorted by the rescorer. We use these values in the coordinating node to collapse the results of each shard but the rescorer in Lucene cannot access them:
https://github.com/elastic/elasticsearch/blob/99f88f15c5febbca2d13b5b5fda27b844153bf1a/server/src/main/java/org/apache/lucene/search/grouping/CollapseTopFieldDocs.java#L41
I am really sorry I missed that but since it would require a rewriting of the rescorer in Lucene and that the collapsing code is only in es I don't think it is worth the effort.

**Comment by rpedela** [comment]
Doesn't Solr support collapse + rescore (rerank)? The claim that Lucene's rescorer needs a rewrite seems dubious.

**Comment by jimczi** [comment]
I agree that we should be able to rescore collapsed documents but this is more high hanging fruit than I thought which is why I reverted and closed the issue for now (sorry @fred84 ).
The current design of the collapsing in es is not compatible with the rescorer and will require some internal refactoring. I've started to work on this refactoring and when it's ready we'll reevaluate this pr if @fred84 still wants to work on it ;).

**Comment by fred84** [comment]
@jimczi let me now when I can start this issue again :)

**Comment by jimczi** [comment]
Sorry it took me some time to come back at this. I checked why Solr was able to rescore the collapsed documents seamlessly and found out that they force the routing of each group in a single shard. This means that all the documents belonging to a single group are on the same shard so the rescoring is always done on the final head of the group. In es we don't enforce the routing so each group can be spread over multiple shards. This complicates the rescoring since it is always applied at the shard level and in this case on the temporary head of the groups (we don't know the final head in the shard since another shard can contain a better document for that group). For this reason I am reluctant to add this functionality because it might be surprising to see a head in a group that is not the best document of that group in the final response. This can happen if the rescoring gives a score to a document in a shard that is better that the score of the best document in the group which is in another shard. I don't see how we could avoid this unless we force the routing of the groups.

**Comment by damitkwr** [comment]
Is there a way to revisit this functionality? This limitation curtails the use of any LTR algorithms

**Comment by DajunZhou** [comment]
Excuse me, is there a way to apply collapsing before rescoring？

**Comment by DajunZhou** [comment]
Excuse me, is there a way to apply  rescoring before collapsing？

**Comment by lasagar** [comment]
> Is there a way to revisit this functionality? This limitation curtails the use of any LTR algorithms

I am looking for this option for LTR rescoring

**Comment by hosuaby** [comment]
Hello, @jimczi 

I am looking for possible (even limited) solution to make collapse working with rescore (especially LTR).
For the moment, I created a small plugin, that support search extension:
```json
"ext": {
    "post-rescore" : { /* ... */ }
}
```

This `post-rescore` has the same syntax as original rescore, but it is executed as `FetchSubPhase` (so I assume, that it's executed on coordinating node).

The present solution rescore only results present on current page.

Code of solution is here:
https://github.com/hosuaby/bug-elasticsearch-collapse-rescore/tree/master/src/main/java/io/hosuaby/elasticsearch

I will be thankful for any feedback and suggestions.

**Comment by Morikko** [comment]
@hosuaby you might have a look to this discussion: https://github.com/elastic/elasticsearch/issues/60946 it looks close to your proposition

**Pull request merged**